In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import sklearn

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression     # 분류 알고리즘
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score,  confusion_matrix, classification_report
# sklearn.metrics : 평가가 잘 되었는지 확인하는 용도

In [284]:
data = load_breast_cancer()
print( type(data))
print(dir(data))
#print( data.head() )       # dataFrame이 아니라서 head()는 실행이 안됨

<class 'sklearn.utils._bunch.Bunch'>
['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [285]:
X = data.data       # 훈련에 사용할 데이터를 가져옴         # 훈련데이터는 일반적으로 대문자를 사용함. (대문자 X)
print(X)            # 평균이 0이고, 표준편차가 1인 정규화시킨 데이터를 가져옴.

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


In [286]:
y = data.target       # 타겟 값은 소문자 y로 가져옴.
print(y)              # target은 실제 정답 값.

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

In [287]:
print(data.DESCR)

.. _breast_cancer_dataset:

Breast cancer Wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

:Number of Instances: 569

:Number of Attributes: 30 numeric, predictive attributes and the class

:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)
    - texture (standard deviation of gray-scale values)
    - perimeter
    - area
    - smoothness (local variation in radius lengths)
    - compactness (perimeter^2 / area - 1.0)
    - concavity (severity of concave portions of the contour)
    - concave points (number of concave portions of the contour)
    - symmetry
    - fractal dimension ("coastline approximation" - 1)

    The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.

    - 

In [288]:
print(X.shape)          # 앞이 데이터 갯수, 뒤에가 열. (569개)
print(len(y))           # (569, 30) : 30개의 피쳐와 실제 데이터는 569개이다

(569, 30)
569


In [289]:
print(y.shape)

(569,)


# 실제 훈련을 진행하기 전에, 훈련데이터와 검증 데이터를 미리 나눠야함.

### supervisor (지도학습) 으로 진행

In [290]:
# 데이터 분할(학습 80%, 테스트 20% )                # random_state=42 : 실행때마다 똑같은 값으로 나오게함.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)        # 데이터 값과 정답 값, size 를 넣어준다. (size 기본 값은 0.25)

# 처음에 7:3으로 나누고, 그 다음에 3 안에서 5:5로 나눈다. 
# 훈련할 데이터는 일반적으로 대문자로 작성한다.
# 값, 타겟은 일반적으로 소문자로 작성한다.

In [291]:
print(X_train.shape, y_train.shape)         # 데이터의 갯수가 같아야함.

(455, 30) (455,)


In [292]:
print(X_test.shape, y_test.shape)

(114, 30) (114,)


# 1. LogisticRegression  분류 알고리즘

LogisticRegression( )알고리즘의 penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='deprecated', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None  


1. **penalty:**  
정규화 방법을 지정한다. 'l2'는 L2 정규화를 의미하며, 이는 가중치의 제곱합을 최소화하여 과적합을 방지하는 데 도움을 준다. 'l1'을 선택하면 L1 정규화를 사용하여 일부 가중치를 0으로 만들 수 있다. 'elasticnet'을 선택하면 L1과 L2 정규화를 혼합하여 사용할 수 있다.  
**2. dual:**  
이진 분류 문제에서 이중 문제를 해결할지를 결정한다. True로 설정하면 이중 문제를 해결하고, False로 설정하면 원래의 문제를 해결한다. 일반적으로 이진 분류에서는 False로 설정하는 것이 더 효율적이다.  
**3. tol:**  
수렴 기준을 설정합니다. 알고리즘이 수렴했다고 판단하는 기준으로, 기본값은 0.0001이다. 이 값이 작을수록 더 정밀한 수렴을 요구한다.  
**4. C:**  
정규화 강도를 조절하는 하이퍼파라미터이다. 값이 작을수록 정규화가 강해지고, 값이 클수록 정규화가 약해진다. 기본값은 1.0이다.  
**5. fit_intercept:**  
절편(intercept)을 모델에 포함할지를 결정한다. True로 설정하면 절편을 포함하고, False로 설정하면 절편을 포함하지 않는다. 기본값은 True이다.  
**6. intercept_scaling:**  
다중 클래스 문제에서 절편을 조정하는 데 사용된다. 기본값은 1이며, 이 값은 절편의 스케일을 조정하는 데 사용된다.  
**7. class_weight:**  
클래스의 가중치를 설정한다. None으로 설정하면 모든 클래스의 가중치가 동일하게 설정됩니다. 클래스 불균형이 있는 경우, 특정 클래스에 더 높은 가중치를 부여하여 모델이 해당 클래스를 더 잘 학습하도록 할 수 있다.  
**8. random_state:**  
난수 생성기의 시드를 설정합니다. 동일한 시드를 사용하면 동일한 결과를 재현할 수 있다. 기본값은 None이다.  
**9. solver:**  
최적화 알고리즘을 선택한다. 'lbfgs', 'newton-cg', 'liblinear', 'sag', 'saga' 중 하나를 선택할 수 있다. 각 솔버는 서로 다른 특성과 성능을 가지므로 데이터셋에 따라 적절한 솔버를 선택하는 것이 중요하다.  
**10. max_iter:**      
알고리즘의 최대 반복 횟수를 설정한다. 기본값은 100이며, 이 값에 도달하기 전에 수렴하면 알고리즘이 종료된다.  
**11. multi_class:**  
다중 클래스 문제에서의 처리 방법을 설정한다. 'deprecated'는 더 이상 사용되지 않는 옵션이며, 대신 'ovr' (One-vs-Rest) 또는 'multinomial'을 사용할 수 있다.  
**12. verbose:**  
출력의 상세 정도를 설정한다. 0이면 출력이 없고, 1 이상의 값으로 설정하면 더 많은 정보가 출력된다.  
**13. warm_start:**  
이전에 학습한 모델을 재사용할지를 결정한다. True로 설정하면 이전 모델의 가중치를 초기값으로 사용하여 학습을 계속할 수 있다.  
**14. n_jobs:**  
병렬 처리를 위한 작업 수를 설정한다. None으로 설정하면 기본값이 사용되며, -1로 설정하면 모든 CPU 코어를 사용한다.  
**15. l1_ratio:**  
L1과 L2 정규화의 비율을 설정한다. 'elasticnet'을 사용할 때만 의미가 있으며, 기본값은 None이다.

## 오류 발생

In [293]:
# 모델 학습(로지스틱 회귀)
model = LogisticRegression(max_iter=10000)      # 분류를 위한, 분류 알고리즘 사용       #max_iter 지정 필수. But, 안에 숫자값은 의미가 없음.
model.fit(X_train, y_train)
print(dir(model))

['C', '__annotations__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__sklearn_clone__', '__sklearn_tags__', '__str__', '__subclasshook__', '__weakref__', '_doc_link_module', '_doc_link_template', '_doc_link_url_param_generator', '_get_class_level_metadata_request_values', '_get_doc_link', '_get_fitted_attr_html', '_get_metadata_request', '_get_param_names', '_get_params_html', '_html_repr', '_init_callback_context', '_parameter_constraints', '_predict_proba_lr', '_repr_html_inner', '_repr_mimebundle_', '_validate_params', 'class_weight', 'classes_', 'coef_', 'decision_function', 'densify', 'dual', 'fit', 'fit_intercept', 'get_metadata_routing', 'get_params', 'intercept_', 'intercept_scaling', 'l1_ratio'

In [294]:
import platform
print(platform.system() )

# html 다이어그램 출력 기능을 끄고 텍스트로 표시
if platform.system() == 'Windows' :
    sklearn.set_config(display='text')


# 모델 학습 (로지스틱 회귀)
model = LogisticRegression(max_iter=10000)
model.fit(X_train, y_train)

Windows


LogisticRegression(max_iter=10000)

In [295]:
# 예측 (클래스 에측 및 확률 예측)
y_pred = model.predict(X_test)
print(y_pred)
print(y_pred == y_test )

[1 0 0 1 1 0 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 1 0 1 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 1 1 1 0 0 1 1 1 0 0 1 1 0 0 1 0
 1 1 1 1 1 1 0 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 1 0 0]
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True False  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True False  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True False  True  True  True  True False  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True False  True]


In [296]:
print( pd.Series(y_pred == y_test).unique() )

[ True False]


In [297]:
print( pd.Series( y_pred == y_test).value_counts() )

True     109
False      5
Name: count, dtype: int64


In [298]:
y_prob = model.predict_proba(X_test)
print(y_prob)

[[1.21238266e-01 8.78761734e-01]
 [9.99999970e-01 3.03624981e-08]
 [9.98405090e-01 1.59490988e-03]
 [1.21353825e-03 9.98786462e-01]
 [1.39796565e-04 9.99860203e-01]
 [1.00000000e+00 1.94458720e-10]
 [1.00000000e+00 6.09483521e-11]
 [9.80773102e-01 1.92268979e-02]
 [1.58520339e-02 9.84147966e-01]
 [5.63029459e-03 9.94369705e-01]
 [7.04321239e-02 9.29567876e-01]
 [9.99193285e-01 8.06714616e-04]
 [1.10171295e-02 9.88982871e-01]
 [8.24174930e-01 1.75825070e-01]
 [7.79253365e-04 9.99220747e-01]
 [9.98272428e-01 1.72757217e-03]
 [1.18193344e-03 9.98818067e-01]
 [1.18481436e-05 9.99988152e-01]
 [5.99658938e-07 9.99999400e-01]
 [9.99999288e-01 7.11997904e-07]
 [1.71640882e-01 8.28359118e-01]
 [7.63965178e-03 9.92360348e-01]
 [9.99999994e-01 6.05567114e-09]
 [3.32835881e-04 9.99667164e-01]
 [3.43832406e-03 9.96561676e-01]
 [2.38455951e-04 9.99761544e-01]
 [1.01380772e-03 9.98986192e-01]
 [1.01253871e-03 9.98987461e-01]
 [4.65847539e-03 9.95341525e-01]
 [9.99999993e-01 7.24261984e-09]
 [1.452603

In [299]:
# 출력 : 혼합행렬
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[39  4]
 [ 1 70]]


출력 :
[[39  4]
 [ 1 70]]

 행이 실제,
 열이 예측


                         예측 
              음성[0]   양성[1] 
실   음성[0]  [[ TN  /   FP  ]  
제   양성[1]  [  FN  /   TP ]]

In [300]:
# 평가지표 계산

# y_test: 정답,  y_pred: 예측
acc = accuracy_score(y_test, y_pred)        # 정확도     전체 중에서 실제 정답인 비율
prec = precision_score(y_test, y_pred)      # 정밀도     Positive 예측에서 진짜 Postive인 경우
rec = recall_score(y_test, y_pred)          # 재현율     실제 Postive인 범위 중에서 Postive라고 예측한 비율
f1 = f1_score(y_test, y_pred)               # F1 Score  정밀도와 재현율을 결합한 지표

print('평가 지표')
print(f'Accuracy: {acc}' )
print(f'Precision: {prec}')
print(f'Recal: {rec}')
print(f'F1 Score: {f1}')

평가 지표
Accuracy: 0.956140350877193
Precision: 0.9459459459459459
Recal: 0.9859154929577465
F1 Score: 0.9655172413793104


In [301]:
target_names = data.target_names  # ['malignant', 'benign']     # 0:악성 , 1:양성
print(target_names)

'''
| 용어            | 한글 해석   | 설명                                                                          |
| -------------   | ------- | --------------------------------------------------------------------------- |
| **malignant**   | 악성 (암)  | 빠르게 자라고 전이할 가능성이 높은 종양이나 질병. 보통 **암(Cancer)** 을 뜻함. 치료가 어렵고 생명에 위협이 될 수 있음. |
| **benign**      | 양성  (비암성) | 성장하긴 하지만 전이하지 않으며 비교적 **위험하지 않은 상태**. 수술 등으로 완치 가능한 경우가 많음.                 |

- malignant: 악성 종양 (암) → 0 또는 Negative Class
- benign: 양성 종양 (비암성) → 1 또는 Positive Class
'''



# 0. 전체 classification_report
print("Classification Report")
# target_names옵션을 설정하지 않으면 0과 1로 표시된다.
print(classification_report(y_test, y_pred))
print("Classification Report")      # 분류 레포트 출력
print(classification_report(y_test, y_pred, target_names=target_names))

['malignant' 'benign']
Classification Report
              precision    recall  f1-score   support

           0       0.97      0.91      0.94        43
           1       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114

Classification Report
              precision    recall  f1-score   support

   malignant       0.97      0.91      0.94        43
      benign       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



In [302]:
# average = Literal['micro', 'macro', 'samples', 'weighted', 'binary']
#1. average=None (클래스별 성능 개별 출력)
print("average=None (클래스별 출력)")
print("Precision:", precision_score(y_test, y_pred, average=None))
print("Recall:", recall_score(y_test, y_pred, average=None))
print("F1 Score:", f1_score(y_test, y_pred, average=None))
print()

# average='micro'
print("average='micro'")
print("Precision:", precision_score(y_test, y_pred, average='micro'))
print("Recall   :", recall_score(y_test, y_pred, average='micro'))
print("F1 Score :", f1_score(y_test, y_pred, average='micro'))
print()

# average='macro'
print("average='macro'")
print("Precision:", precision_score(y_test, y_pred, average='macro'))
print("Recall   :", recall_score(y_test, y_pred, average='macro'))
print("F1 Score :", f1_score(y_test, y_pred, average='macro'))
print()

# average='weighted'
print("average='weighted'")
print("Precision:", precision_score(y_test, y_pred, average='weighted'))
print("Recall   :", recall_score(y_test, y_pred, average='weighted'))
print("F1 Score :", f1_score(y_test, y_pred, average='weighted'))
print()

average=None (클래스별 출력)
Precision: [0.975      0.94594595]
Recall: [0.90697674 0.98591549]
F1 Score: [0.93975904 0.96551724]

average='micro'
Precision: 0.956140350877193
Recall   : 0.956140350877193
F1 Score : 0.956140350877193

average='macro'
Precision: 0.960472972972973
Recall   : 0.9464461185718964
F1 Score : 0.9526381387619444

average='weighted'
Precision: 0.9569049312470365
Recall   : 0.956140350877193
F1 Score : 0.9558014271241044



## 데이터가 불균형일 때 발생되는 문제

# 불균형한 이진 분류로 만들기 위해 숫자 '7'만 True 나머지 숫자는 False로 설정

- 마찬가지로 fit은 아무것도 하지 않음
- predict에서는 무조건 0을 반환함
- 모든 예측을 다 0으로 하는 것! (모두 7이 아닌 다른 숫자라고 예측)
- 당연히 모델로서 말이 안 되는데 이것의 정확도를 측정

In [303]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
import numpy as np
import pandas as pd

In [304]:
class MyFakeClassifier(BaseEstimator):
    def fit(self,X,y):                                                      # 훈련 메소드
        pass

    # 입력값으로 들어오는 X 데이터 셋의 크기만큼 모두 0값으로 만들어서 반환
    def predict(self,X):                                                    # 예측 메소드
        return np.zeros( (len(X), 1) , dtype=bool)

digits = load_digits()
print('### digits.data.shape:', digits.data.shape)
print(digits.target)
print('### digits.target.shape', digits.target.shape)

### digits.data.shape: (1797, 64)
[0 1 2 ... 8 9 8]
### digits.target.shape (1797,)


In [305]:
# 일반적으로 분류 알고리즘은 뒤에 이름이 'Classifier'로 붙는다
# regression만 예외였던것임.

In [306]:
print(digits.target )

# 조건이 만족하면 1로 하고, 나머진 0으로 하겠다.    =>   1과 0의 불균형 데이터를 만들려는 목표.

print(digits.target_names )
y = (digits.target==7).astype(int)
print(y)


print(len(y))
print(pd.Series(y).value_counts())

[0 1 2 ... 8 9 8]
[0 1 2 3 4 5 6 7 8 9]
[0 0 0 ... 0 0 0]
1797
0    1618
1     179
Name: count, dtype: int64


In [307]:
X_train, X_test, y_train, y_test = train_test_split(digits.data, y, random_state=42)
print(pd.Series(y_train).value_counts())

0    1209
1     138
Name: count, dtype: int64


In [308]:
print(pd.Series(y_test).value_counts())

0    409
1     41
Name: count, dtype: int64


In [309]:
fakeclf = MyFakeClassifier()
fakeclf.fit(X_train, y_train)

fakepred = fakeclf.predict(X_test)

print('모든 예측을 0으로 하여도 정확도는 :{:.3f}' .format(accuracy_score(y_test, fakepred) ))

cf = confusion_matrix(y_test, fakepred)
print(cf)

모든 예측을 0으로 하여도 정확도는 :0.909
[[409   0]
 [ 41   0]]


# (위) 데이터 불균형이 발생되면 정확도가 높게 나올 수 있는  문제가 있다.

In [310]:
f1 = f1_score(y_test, fakepred)
print(f'f1_score: {f1}')

f1_score: 0.0


In [311]:
prec = precision_score(y_test, fakepred)
print(f' precision : {prec}')

 precision : 0.0


c:\sk-encoa\data-collection-workspace\.venv\Lib\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


- 모든 예측을 0으로 했지만, 0 (7이 아닌 숫자)이 1 (숫자 7)보다 훨씬 많은 불균형 데이터였기 때문에, 정확도가 무려 90%가 나왔다.

- 이렇게 두 가지 예제로 불균형 데이터에서 정확도의 성능 평가지표로서 한계를 보였다

In [312]:
# # 1. average = None (클래스별 개별) default는 binary

# print('average = None (클래스별 출력)')
# print('Precision : ', precision_score(y_test, y_pred, average=None))
# print('Recall : ', recall_score(y_test, y_pred, average=None) )

## 결과 값 :

average=None (클래스별 출력)
Precision: [0.975      0.94594595]
Recall: [0.90697674 0.98591549]
F1 Score: [0.93975904 0.96551724]

average='micro'
Precision: 0.956140350877193
Recall   : 0.956140350877193
F1 Score : 0.956140350877193

average='macro'
Precision: 0.960472972972973
Recall   : 0.9464461185718964
F1 Score : 0.9526381387619444

average='weighted'
Precision: 0.9569049312470365
Recall   : 0.956140350877193
F1 Score : 0.9558014271241044

0. average='binary' (default)
   - positive label을 1로 간주하고 계산한다.
   - y_true에 0과 1 두 개 클래스만 있어야 한다.

1. average=None (클래스별 성능 개별 출력)

| 클래스               | Precision | Recall | F1-score | Support |
| ----------------- | --------- | ------ | -------- | ------- |
| **0 (malignant)** | 0.97      | 0.91   | 0.94     | 43      |
| **1 (benign)**    | 0.95      | 0.99   | 0.97     | 71      |

- 각 클래스별로 정밀도, 재현율, F1-score를 개별적으로 계산하여 보여준다.
- 모델이 클래스 0 (악성 종양)을 상대적으로 덜 잘 맞추고 있다는 것을 파악할 수 있다.

2. average='micro' (전체 기준 - 샘플 단위의 합)  
- micro 평균은 모든 TP, FP, FN을 전 클래스 기준으로 합산한 뒤 계산한다.
- 이 경우 정확도(Accuracy) = Precision = Recall = F1-score가 된다.
- 총 정답 수: TP(전체 맞춤 수) = 0 클래스에서 39개 + 1 클래스에서 70개 = 109
- 전체 샘플 수: 114
- 따라서:    Accuracy = Precision = Recall = F1 = 109 / 114 ≈ 0.9561  

3. average='macro' (단순 평균)  
- 클래스별 점수를 동일한 비중으로 평균함
  Macro Precision = (0.97 + 0.95) / 2 = 0.96  
  Macro Recall    = (0.91 + 0.99) / 2 = 0.95  
  Macro F1-score  = (0.94 + 0.97) / 2 = 0.955 ≈ 0.95  

- 클래스 0과 1의 비중이 같다고 가정하고 평균 내기 때문에, 클래스 불균형이 있는 경우에는 왜곡이 있을 수 있다.

4. average='weighted' (샘플 수 기반 가중 평균)  
-  각 클래스의 점수에 해당 클래스의 샘플 개수(Support)를 곱해서 평균
  Weighted Precision = (0.97×43 + 0.95×71) / 114 ≈ 0.958    
  Weighted Recall    = (0.91×43 + 0.99×71) / 114 ≈ 0.961  
  Weighted F1-score  = (0.94×43 + 0.97×71) / 114 ≈ 0.961  

- 실제 출력값은 모두 반올림되어 0.96으로 나옴

- 데이터 불균형을 고려한 평균이므로, 실전에서 가장 신뢰도 있는 지표 중 하나임

5. average 옵션 요약

| 평균 방식      | 설명                    | 특징               | 언제 사용하면 좋은가?                 |
| ---------- | --------------------- |------------ | ---------------------------- |
| `None`     | 클래스별 개별 지표 확인         |  `array` 형태       | 특정 클래스 성능 비교 시               |
| `micro`    | 전체 샘플 기준으로 정밀도/재현율 계산 | 불균형 데이터에 취약      | 다중 클래스, 불균형 무시하고 전체 성능 확인 시  |
| `macro`    | 클래스별 단순 평균            | 클래스 중요도 균일할 때 사용 |클래스 간 균형 평가 시                |
| `weighted` | 클래스별 가중 평균            | 불균형 데이터에 적합      |**불균형 데이터에서의 전체 성능 평가** 시 유리 |